### PR TITLE
[SPARK-9172][SQL] Make DecimalPrecision support for Intersect and Except

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-9172

Simply make `DecimalPrecision` support for `Intersect` and `Except` in addition to `Union`.

Besides, add unit test for `DecimalPrecision` as well.

### CODE REVIEW

**Comment by viirya** [comment]
Wait for #7510 to solve the compilation error.

**Comment by liancheng** [comment]
retest this please

**Comment by rxin** [comment]
cc @cloud-fan for review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 1]
```diff
@@ -362,47 +362,59 @@ object HiveTypeCoercion {
...
-          case other => other
-        }
+    private def castDecimalPrecision(
+        left: LogicalPlan,
+        right: LogicalPlan): (LogicalPlan, LogicalPlan) = {
+      val castedInput = left.output.zip(right.output).map {
+        case (lhs, rhs) if lhs.dataType != rhs.dataType =>
+          (lhs.dataType, rhs.dataType) match {
+            case (DecimalType.Fixed(p1, s1), DecimalType.Fixed(p2, s2)) =>
+              // Union decimals with precision/scale p1/s2 and p2/s2  will be promoted to
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: nit: update the `Union` in comments.

**Comment by cloud-fan** [comment]
lgtm

**Comment by rxin** [comment]
Thanks - merging this in.

**Pull request closed without merging**